### PR TITLE
docs: ScrapeConfig proposal

### PR DESCRIPTION
**Pull request by xiu**
## Description
This is a first shot at a proposal for ScrapeConfig, opening it early to take feedback in.

Related tickets:
* #2787 
* #3447

## Type of change

_What type of changes does your code introduce to the Prometheus operator? Put an `x` in the box that apply._

- [ ] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [ ] `FEATURE` (non-breaking change which adds functionality)
- [ ] `BUGFIX` (non-breaking change which fixes an issue)
- [ ] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [x] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

## Changelog entry

```release-note

```

### CODE REVIEW

**Review by JoaoBraveCoding — COMMENTED** [review]
Disclaimer: I'm not a maintainer so take it with a big pinch of salt ;) 

Great start! I left some initial comments on things that popped up in my mind while reading it. Some points are more for the future, I was not sure if they should already be included here or left out for follow-up discussions, so I added them regardless.

Thanks a lot for putting it together @xiu 🙏

File: Documentation/proposals/202212-scrape-config.md
Comment (xiu) [reply Documentation/proposals/202212-scrape-config.md:14]: Thanks! What do you think about:
> prometheus-operator misses a way to scrape external targets using CRD. Users have either been abusing the `Probe` CRD (#3447) or `additionalScrapeConfig` to do so. 
> We currently have several CRDs generating scrape configurations. With this proposition of creating a new `ScrapeConfig` CRD, we also have an interest in consolidating the logic to generate scrape configuration in one place by having all the other CRDs generate `ScrapeConfig` resources instead.

File: Documentation/proposals/202212-scrape-config.md
Comment (xiu) [reply Documentation/proposals/202212-scrape-config.md:18]: Do you have a reference to an issue? I'd love to have a bit more context on this one to make sure I understand it properly.

File: Documentation/proposals/202212-scrape-config.md
Comment (xiu) [reply Documentation/proposals/202212-scrape-config.md:23]: What do you think of:
> # Goals
> * Provide a way for users to self-service adding scrape targets
> * Consolidate the scrape configuration generation logic in a central point for other resources to use

File: Documentation/proposals/202212-scrape-config.md
Comment (JoaoBraveCoding) [reply Documentation/proposals/202212-scrape-config.md:18]: I took that from the first issue you have linked, it's in the first 3 comments https://github.com/prometheus-operator/prometheus-operator/issues/2787

File: Documentation/proposals/202212-scrape-config.md
Comment (JoaoBraveCoding) [reply Documentation/proposals/202212-scrape-config.md:14]: Changed it slightly but feel free to use which one of the texts looks better to you :) 

> prometheus-operator misses a way to scrape external targets using CRD. Users have either been abusing the Probe CRD (https://github.com/prometheus-operator/prometheus-operator/issues/3447) or additionalScrapeConfig to do so.
Furthermore, currently, there is a lot of code duplication due to the operator supporting several CRDs that generate scrape configurations. With the new ScrapeConfig CRD, it would be possible to consolidate some of that logic, where the other `*Monitor` CRDs could be migrated so that they create a ScrapeConfig resource that would ultimately be used by the operator to generate scrape configuration.

File: Documentation/proposals/202212-scrape-config.md
Comment (JoaoBraveCoding) [reply Documentation/proposals/202212-scrape-config.md:23]: Sounds good 👍

File: Documentation/proposals/202212-scrape-config.md
Comment (xiu) [reply Documentation/proposals/202212-scrape-config.md:18]: Had a chat with @JoaoBraveCoding: this usecase needs `kubernetes_sd_config` to work with `ScrapeConfig`, this is out of scope for this first version but we should be able to add it easily once the CRD is in place. The alternative for this usecase is still to use `additionalScrapeConfigs`.

File: Documentation/proposals/202212-scrape-config.md
Comment (xiu) [reply Documentation/proposals/202212-scrape-config.md:35]: Done, let me know what you think, it's pretty simple and potentially naive.

File: Documentation/proposals/202212-scrape-config.md
Comment (JoaoBraveCoding) [inline Documentation/proposals/202212-scrape-config.md:49]: I was more thinking about it in golang 😅 but up to you

```golang
// Scrape Config defines a prometheus scrape_config.
type ScrapeConfig struct {
	metav1.TypeMeta   `json:",inline"`
	metav1.ObjectMeta `json:"metadata,omitempty"`
	// Specification of the desired scrape configuration More info:
	// https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#spec-and-status
	Spec ScrapeConfigSpec `json:"spec"`
}

// ScrapeConfigSpec is a specification of the desired behavior of the ScrapeConfig. More info:
// https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#spec-and-status
// +k8s:openapi-gen=true
type ScrapeConfigSpec struct {
         // The list of static configs
        StaticConfigs  []StaticConfig `json:"staticConfigs,omitempty"`
         // The list of file service discovery configs
        FileSDConfigs []SecretOrConfigMap `json:"fileSDConfig,omitempty"`
        // The list of relabel configurations.
        RelabelConfigs []RelableConfig `json:"reliableConfig,omitempty"`
}

// +k8s:openapi-gen=true
type StaticConfig {
         Targets []string `json:"targets"`
         Labels map[string]string `json:"labels,omitempty"`
}

// Already exists
type SecretOrConfigMap 

// Already exists
type RelabelConfigs
```

I'm not sure if I'm missing something

**Review by JoaoBraveCoding — COMMENTED** [review]
I tried boostraping the CRD definition in golang so we can agree already on what it will look like in the code

File: Documentation/proposals/202212-scrape-config.md
Comment (xiu) [reply Documentation/proposals/202212-scrape-config.md:49]: Right, I didn't want to be too strict in the definition as that definition might change as we go about implementing it. @simonpasquier mentioned something about not having the definition in the proposal in the contributor office hours.

**Review by slashpai — COMMENTED** [review]
A few nits

File: Documentation/proposals/202212-scrape-config.md
Comment (simonpasquier) [reply Documentation/proposals/202212-scrape-config.md:54]: I'm fine with a raw YAML version here. The details can be discussed in pull requests.

File: Documentation/proposals/202212-scrape-config.md
Comment (simonpasquier) [inline Documentation/proposals/202212-scrape-config.md:89]: we would need support for kubernetes SD config too?

File: Documentation/proposals/202212-scrape-config.md
Comment (simonpasquier) [inline Documentation/proposals/202212-scrape-config.md:62]: I wonder if we shouldn't also include HTTP SD config from the start. It could be useful to experiment with target sharding techniques such as the OpenTelemetry collector's [target allocator](https://github.com/open-telemetry/opentelemetry-operator#target-allocator).

File: Documentation/proposals/202212-scrape-config.md
Comment (simonpasquier) [inline Documentation/proposals/202212-scrape-config.md:13]: I think that it'd be helpful if you provide more details about your specific use case @xiu 
#2787 was initially about Kubernetes node service discovery which could be mentioned here too. I'll go through the existing issues to see if other use cases are worth mentioning.

**Review by simonpasquier — COMMENTED** [review]
I like it a lot!

File: Documentation/proposals/202212-scrape-config.md
Comment (xiu) [reply Documentation/proposals/202212-scrape-config.md:62]: We can, it shouldn't be too hard and is generic enough to help a lot of people from the start. I would limit to those (static/file/http) on the first step though and add other SDs as we go. This should be decoupled from the refactoring effort. Thoughts?

File: Documentation/proposals/202212-scrape-config.md
Comment (xiu) [reply Documentation/proposals/202212-scrape-config.md:13]: I've added three use cases to the document.

File: Documentation/proposals/202212-scrape-config.md
Comment (xiu) [reply Documentation/proposals/202212-scrape-config.md:89]: See https://github.com/prometheus-operator/prometheus-operator/pull/5279#discussion_r1089306707. I think this should come but maybe not as part of the first version. Once we have the foundation, we can onboard the other CRDs (and will need to as we'll transition the ServiceMonitor/Probe logic to it).

File: Documentation/proposals/202212-scrape-config.md
Comment (xiu) [reply Documentation/proposals/202212-scrape-config.md:87]: I've added httpSDConfig to the list

Comment (xiu) [reply Documentation/proposals/202212-scrape-config.md:87]: https://github.com/prometheus-operator/prometheus-operator/pull/5279/files#diff-a6321575a64799dcf97cfe7e62ce050189f50e6cd33fc6cc495f118f6cfefcddR143-R144

**Review by ArthurSens — COMMENTED** [review]
Hi @xiu, thanks for writing this up!

During my reading one single question came up: How would the operator attach a ScrapeConfig resource with a Prometheus resource?

I assume a Prometheus CRD can only have a single ScrapeConfig resource attached, right? What would happen if a Prometheus is configured to use ScrapeConfig and several ServiceMonitors at the same time for example?

**Comment by xiu** [comment]
> Hi @xiu, thanks for writing this up!
> 
> During my reading one single question came up: How would the operator attach a ScrapeConfig resource with a Prometheus resource?
> 
> I assume a Prometheus CRD can only have a single ScrapeConfig resource attached, right? What would happen if a Prometheus is configured to use ScrapeConfig and several ServiceMonitors at the same time for example?

In this case, the aim is to have multiple `ScrapeConfig` resources and they would be used to build the configuration the same way as a `ServiceMonitor` for instance. One goal of this proposal is to allow users to self service scrape configurations. For instance TeamA could define their own specific scrape configurations without conflicting with TeamB and removing the bottleneck that is `additionalScrapeConfigurations`.

**Comment by ArthurSens** [comment]
> In this case, the aim is to have multiple `ScrapeConfig` resources and they would be used to build the configuration the same way as a `ServiceMonitor` for instance. One goal of this proposal is to allow users to self service scrape configurations. For instance TeamA could define their own specific scrape configurations without conflicting with TeamB and removing the bottleneck that is `additionalScrapeConfigurations`.

Oh I see, I'm not sure if I missed it somewhere, but would it make sense to add to the doc that they are not mutual exclusive?

File: Documentation/proposals/202212-scrape-config.md
Comment (xiu) [inline Documentation/proposals/202212-scrape-config.md:69]: @ArthurSens I would image the best place to add that information is here. Currently we say that the `ScrapeConfig` will act the same as the others and append scrape configurations to the configuration. Should I rephrase to the following?
> [...] we will create a new `ScrapeConfig` CRD, this new CRD will act the same as the other CRDs and append scrape configurations to the configuration. Usage of `ScrapeConfig` doesn't exclude the use of the other CRDs, they are not mutually exclusive.

**Comment by xiu** [comment]
> > In this case, the aim is to have multiple `ScrapeConfig` resources and they would be used to build the configuration the same way as a `ServiceMonitor` for instance. One goal of this proposal is to allow users to self service scrape configurations. For instance TeamA could define their own specific scrape configurations without conflicting with TeamB and removing the bottleneck that is `additionalScrapeConfigurations`.
> 
> Oh I see, I'm not sure if I missed it somewhere, but would it make sense to add to the doc that they are not mutual exclusive?

resolved here: https://github.com/prometheus-operator/prometheus-operator/pull/5279#discussion_r1090404507

**Review by JoaoBraveCoding — APPROVED** [review]
After reviewing the latest version, from my side it's a lgtm 👍 Thanks for the contribution @xiu! 🚀

File: Documentation/proposals/202212-scrape-config.md
Comment (simonpasquier) [inline Documentation/proposals/202212-scrape-config.md:125]:
```suggestion
refreshInterval: 60s
```

File: Documentation/proposals/202212-scrape-config.md
Comment (simonpasquier) [inline Documentation/proposals/202212-scrape-config.md:113]: After giving it more thoughts, I'm wondering if file SD configuration is really needed and if it brings enough value compared to static targets? The user experience with file SD would be:
* create the configmaps/secrets for targets
* create the scrape config object
* update the configmaps/secrets
With the static config, the user has to deal only with the scrape config object which feels less burden and more idiomatic. There's also the challenge of syncing the user's secrets/configmaps to an operator-owned secret as well as avoiding the redeployment of Prometheus on secret addition/removal (similar to #5085). Finally it's less obvious how the operator reports that the user's secret/configmap is invalid.

From a classic Prometheus user's standpoint, file SD is interesting because compared to static config, Prometheus can pick up changes almost immediately. But in the operator's case, it's more an implementation detail and it might be possible to "translate" static targets into a file SD configuration?

Comment (paulfantom) [reply Documentation/proposals/202212-scrape-config.md:113]: I was also wondering about this. @xiu afair you mentioned that you have a use case for file sd with the operator, could you share more?

Comment (xiu) [reply Documentation/proposals/202212-scrape-config.md:113]: Sure thing. I added file_sd as I'm mostly interested in this one. Our usecase is that we have a job pulling discovery data from internal systems and landing it in a configmap regularly like:
```json
[
  {
    "labels": {
      "rack": "1",
      "cluster": "clusterA",
      "type": "typeA",
      "team": "teamB"
    },
    "targets": [
      "host1.fqdn"
    ]
  },
  {
    "labels": {
      "rack": "2",
      "cluster": "clusterB",
      "type": "typeB",
      "team": "teamB"
    },
    "targets": [
      "host2.fqdn"
    ]
  }
]
```

With this configmap mounted in Prometheus, we can then create scraping rules like this, that we use `additionalScrapeConfigs` for:
```yaml
- file_sd_configs:
  - files:
    - ../configmaps/inventory/file.json
  job_name: my_job
  relabel_configs:
    - action: keep
      regex: typeA
      source_labels:
      - type
    - regex: (.*)
      replacement: ${1}:9100
      source_labels:
      - __address__
      target_label: __address__
  static_configs: []
```

The UX is not great of course, and we'll be working at revamping that, but it is the system we have today.

In general though, I would imagine `ScrapeConfig` to stick with Prometheus' primitives and have other higher level CRDs (`ServiceMonitor`/`PodMonitor`) be more opinionated. I'm happy to jump on a call tomorrow to describe my usecase in more details if needed.

Comment (paulfantom) [reply Documentation/proposals/202212-scrape-config.md:113]: I think that's a valid use-case and others might have a similar one. However, it somehow feels like rebranded `additionalScrapeConfigs` :thinking: Do we want to also have validation on files listed in a ConfigMap?

There is also slight implementation detail we need to remember in this particular case - since ConfigMaps/Secrets are double linked when mounting in a container, we cannot rely on Prometheus just picking up newly changed files from those ConfigMap/Secrets. This means either config-reloader sidecar would need to be updated every time a new `file_sd_config` is added (which results in prometheus redeploy) or the operator would need to somehow track changes in those ConfigMaps (annotation with a checksum?) or some other way of ensuring this. All this means we are adding quite a ton of complexity to the initial implementation and without this particular feature I don't think file_sd would be complete.

Comment (xiu) [reply Documentation/proposals/202212-scrape-config.md:113]: > However, it somehow feels like rebranded additionalScrapeConfigs 🤔
It's one of the goals imo, provide Kubernetes primitives rather than a centralized `additionalScrapeConfigs`.

> Do we want to also have validation on files listed in a ConfigMap?
I think checking that the file exist would already be good. I wouldn't consider validation of the content as needed, at least in the first version. Worst case, we won't discover scrape targets.

> There is also slight implementation detail we need to remember in this particular case - since ConfigMaps/Secrets are double linked when mounting in a container, we cannot rely on Prometheus just picking up newly changed files from those ConfigMap/Secrets.
I'm not sure I get this one though I'm far from the most knowledgeable on the code base 😄 Currently we use one of these configmaps without any issue and it gets updated regularly. Prometheus has [`refresh_interval` on `file_sd_configs`](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#file_sd_config) set to 5m by default to reload the file. Am I missing something?

Comment (paulfantom) [reply Documentation/proposals/202212-scrape-config.md:113]: All good and thank you for the explanation. I may have overthinked it :)

Comment (simonpasquier) [reply Documentation/proposals/202212-scrape-config.md:113]: > I think checking that the file exist would already be good. I wouldn't consider validation of the content as needed, at least in the first version. Worst case, we won't discover scrape targets.

If someone crafts an invalid file SD config and the operator passes it as-is to Prometheus, it's still a sub-par experience IMHO.

The area I'm not clear about is how the operator will synthesize the file SD configs from all the ScrapeConfigs and how they will be passed down to the statefulset. I'll imagine that the operator will copy the secret/configmap data to a common secret so it avoids redeploying the pods everytime a ScrapeConfig is added/removed/updated? We also have to deal deal with the secret's limit size (1MiB). So I'd like to see a bit more details in the proposal :)

File: Documentation/proposals/202212-scrape-config.md
Comment (simonpasquier) [inline Documentation/proposals/202212-scrape-config.md:118]:
```suggestion
refreshInterval: 5m
```

File: Documentation/proposals/202212-scrape-config.md
Comment (xiu) [reply Documentation/proposals/202212-scrape-config.md:113]: > If someone crafts an invalid file SD config and the operator passes it as-is to Prometheus, it's still a sub-par experience IMHO.

Absolutely, this can come as we iterate. I wouldn't put this as part of the first version, the benefits of the CRD is still very high compared to the current solution.

> The area I'm not clear about is how the operator will synthesize the file SD configs from all the ScrapeConfigs and how they will be passed down to the statefulset. I'll imagine that the operator will copy the secret/configmap data to a common secret so it avoids redeploying the pods everytime a ScrapeConfig is added/removed/updated? We also have to deal deal with the secret's limit size (1MiB). So I'd like to see a bit more details in the proposal :)

The same way we do for `ServiceMonitor`, each `ScrapeConfig` will be passed down to the configuration. The `ScrapeConfig` will not manipulate a configmap nor a secret, the filepath will be a string and it'll be the responsibility of the operator to mount the file via `Prometheus.ConfigMaps`. The operator is not responsible for the SD file.

Comment (xiu) [reply Documentation/proposals/202212-scrape-config.md:113]: I realize the confusion now as I referenced `SecretOrConfigMap` in the proposal:

```yaml
files:
  - <SecretOrConfigMap>[] # https://github.com/prometheus-operator/prometheus-operator/blob/e4e27052f57040f073c6c1e4aedaecaaec77d170/pkg/apis/monitoring/v1/types.go#L1644
refresh_interval: 5m
```

Updating the document to reflect what I just said.

Comment (xiu) [reply Documentation/proposals/202212-scrape-config.md:113]: Done in https://github.com/prometheus-operator/prometheus-operator/pull/5279/commits/9f1cd504392ffd1cda5f05f0c5fe1b64f99dbc3a, @simonpasquier @paulfantom sorry for the confusion there and thanks for the thorough review, let me know what you think with this change :)

File: Documentation/proposals/202212-scrape-config.md
Comment (JoaoBraveCoding) [inline Documentation/proposals/202212-scrape-config.md:92]: nit: just making it consistent 
```suggestion
    - <staticConfig>[] # new resource
```

File: Documentation/proposals/202212-scrape-config.md
Comment (JoaoBraveCoding) [inline Documentation/proposals/202212-scrape-config.md:98]: nit: just making it consistent
```suggestion
    - <relabelConfig>[] # https://github.com/prometheus-operator/prometheus-operator/blob/e4e27052f57040f073c6c1e4aedaecaaec77d170/pkg/apis/monitoring/v1/types.go#L1150
```

**Review by JoaoBraveCoding — COMMENTED** [review]
just some nits I noticed when reading the new discussion

File: Documentation/proposals/202212-scrape-config.md
Comment (simonpasquier) [inline Documentation/proposals/202212-scrape-config.md:100]: I assume that you didn't want to list all the fields but we intend to support all the scrape config fields (apart from the SD mechanisms which are currently out of scope)?

File: Documentation/proposals/202212-scrape-config.md
Comment (xiu) [reply Documentation/proposals/202212-scrape-config.md:97]: I'd rather use the native names but consistency is important, I've updated the proposal with that name.

File: Documentation/proposals/202212-scrape-config.md
Comment (xiu) [reply Documentation/proposals/202212-scrape-config.md:100]: Eventually yes, I leave this to the implementation. For instance, on the `http_sd_config`, we might not have OAuth2.0 right away but it could be implemented at a later stage.

File: Documentation/proposals/202212-scrape-config.md
Comment (xiu) [reply Documentation/proposals/202212-scrape-config.md:70]: Added to the proposal

File: Documentation/proposals/202212-scrape-config.md
Comment (xiu) [reply Documentation/proposals/202212-scrape-config.md:100]: Added more to the proposal in that regard

File: Documentation/proposals/202212-scrape-config.md
Comment (simonpasquier) [inline Documentation/proposals/202212-scrape-config.md:135]:
```suggestion
Also, to help selecting `ScrapeConfig`, a new field will be added to the Prometheus CRD, same as for `ServiceMonitor` , 
`PodMonitor` and `Probe` objects:
```

File: Documentation/proposals/202212-scrape-config.md
Comment (simonpasquier) [reply Documentation/proposals/202212-scrape-config.md:100]: :+1:

File: Documentation/proposals/202212-scrape-config.md
Comment (simonpasquier) [inline Documentation/proposals/202212-scrape-config.md:60]:
```suggestion
  Specifically, it focuses first on `static_configs`, `file_sd_configs`  and `http_sd_configs`.
```

**Review by simonpasquier — APPROVED** [review]
I have a few minor comments but I'm :+1: on the proposal! Great work @xiu!

**Review by paulfantom — APPROVED** [review]
Awesome work! :+1:

**Review by PhilipGough — APPROVED** [review]
@xiu - thanks so much for doing this. Awesome work and I'll merge it prior to the release.

LGTM

**Comment by xiu** [comment]
Thanks everyone! Let's work on the implementation now 🥳

**Pull request merged**
